### PR TITLE
add EditOptions to make edit reusable

### PR DESCRIPTION
**Pull request by adohe-zz**
FYI. Added `EditOptions` to decouple edit implementation from cobra command, thus make edit reusable. Once this get merged, we can refactor `create` commands to put them in their own package. @kubernetes/sig-cli-misc please give a quick review, and I would like to get this merged asap.

### CODE REVIEW

**Comment by fabianofranz** [comment]
@AdoHe is this no longer WIP?

**Comment by pwittrock** [comment]
Also, are the test failures legit?

**Comment by kargakis** [comment]
@AdoHe thanks for working on this! Related issue: https://github.com/kubernetes/kubernetes/issues/38761

**Comment by AdoHe** [comment]
@fabianofranz @pwittrock ready for review, ptal.

**Comment by fabianofranz** [comment]
@k8s-bot verify test this

**Comment by pwittrock** [comment]
@k8s-bot verify test this

File: pkg/kubectl/cmd/util/helpers.go [hunk 1]
```diff
@@ -408,13 +414,22 @@ func AddApplyAnnotationFlags(cmd *cobra.Command) {
 	cmd.Flags().Bool(ApplyAnnotationsFlag, false, "If true, the configuration of current object will be saved in its annotation. This is useful when you want to perform kubectl apply on this object in the future.")
 }
 
+func AddApplyAnnotationVarFlags(cmd *cobra.Command, applyAnnotation *bool) {
```
Comment (pwittrock) [inline pkg/kubectl/cmd/util/helpers.go]: WDYT of creating "Get" versions of these flags as well so they can be used in your code above?

File: pkg/kubectl/cmd/util/editor/editoptions.go [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+			results  = editResults{}
+			original = []byte{}
+			edited   = []byte{}
+			file     string
+		)
+
+		containsError := false
+		var infos []*resource.Info
+		for {
+			switch o.EditMode {
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: Can't this check be moved out of the loop?

File: pkg/kubectl/resource/visitor.go [hunk 3]
```diff
@@ -648,6 +648,16 @@ func RetrieveLazy(info *Info, err error) error {
 	return nil
 }
 
+// CreateAndRefresh creates an object from input info and refreshes info with that object
+func CreateAndRefresh(info *Info) error {
```
Comment (shiywang) [inline pkg/kubectl/resource/visitor.go]: Is this function original move from `pkg/kubectl/cmd/create.go` and made it public ? if so, should we delete the previous one ?

**Comment by pwittrock** [comment]
Removing self to get it out of my list.  Re-add me if you would like me to review it.

**Comment by ymqytw** [comment]
@adohe Are you still working on this? If not, maybe @shiywang can pick this up.
https://github.com/kubernetes/kubernetes/pull/42256 is in progress and may depend on this.

**Comment by adohe** [comment]
Rebased and test cases fixed. @pwittrock @fabianofranz please take a look, I would like to get this merged asap.

File: pkg/kubectl/cmd/util/editor/editoptions.go [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/printers"
+	"k8s.io/kubernetes/pkg/util/crlf"
+
+	"github.com/evanphx/json-patch"
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: Move these imports in between stdlib and kubernetes imports

File: pkg/kubectl/cmd/util/editor/editoptions.go [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+type editPrinterOptions struct {
+	printer   printers.ResourcePrinter
+	ext       string
+	addHeader bool
+}
+
+// Complete completes all the required options
+func (o *EditOptions) Complete(f cmdutil.Factory, out, errOut io.Writer, args []string) error {
+	if o.EditMode != NormalEditMode && o.EditMode != EditBeforeCreateMode {
+		return fmt.Errorf("Unsupported edit mode %q", o.EditMode)
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: errors in fmt.Errorf / errors.New should start lowercased

File: pkg/kubectl/cmd/util/editor/editoptions.go [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	mapper, typer, err := f.UnstructuredObject()
+	if err != nil {
+		return err
+	}
+	b := resource.NewBuilder(mapper, f.CategoryExpander(), typer, resource.ClientMapperFunc(f.UnstructuredClientForMapping), unstructured.UnstructuredJSONScheme)
+	if o.EditMode == NormalEditMode {
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: Can you add a comment here why is this needed for normal edit mode?

File: pkg/kubectl/cmd/util/editor/editoptions.go [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+				return preservedFile(err, file, o.ErrOut)
+			}
+
+			switch o.EditMode {
+			case NormalEditMode:
+				err = visitToPatch(infos, updatedVisitor, o.Mapper, encoder, o.Out, o.ErrOut, &results)
+			case EditBeforeCreateMode:
+				err = visitToCreate(updatedVisitor, o.Mapper, o.Out)
+			default:
+				err = fmt.Errorf("Unsupported edit mode %q", o.EditMode)
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: lowercase

**Comment by shiywang** [comment]
*[pkg/kubectl/cmd/util/editor/editoptions.go, line 349 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39404#-KhKpZHGmYVmvaznXohd:-KhKpZHGmYVmvaznXohe:bvep7dl) ([raw file](https://github.com/kubernetes/kubernetes/blob/7fc7847b0623093e6eef73e818f7adf5dc2575c0/pkg/kubectl/cmd/util/editor/editoptions.go#L349)):*
> ```Go
> 	default:
> 		// if format is not specified, use yaml as default
> 		return &editPrinterOptions{
> ```

I'm not sure if this is right, previously if  specify a invalid ouput, it will prompt  "The flag 'output' must be one of yaml|json" should we keep error output compatible ?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39404)*

**Comment by shiywang** [comment]
just have a quick look, generally lgtm, thanks for this work, I will add a third mode `ApplyEditMode` when this pr got merged : )
cc @ymqytw

File: pkg/kubectl/cmd/edit.go [hunk 8]
```diff
@@ -17,38 +17,17 @@ limitations under the License.
...
 	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
 	"k8s.io/kubernetes/pkg/kubectl/cmd/util/editor"
-	"k8s.io/kubernetes/pkg/kubectl/resource"
 	"k8s.io/kubernetes/pkg/printers"
-	"k8s.io/kubernetes/pkg/util/crlf"
 	"k8s.io/kubernetes/pkg/util/i18n"
 
-	jsonpatch "github.com/evanphx/json-patch"
-	"github.com/golang/glog"
 	"github.com/spf13/cobra"
```
Comment (ymqytw) [inline pkg/kubectl/cmd/edit.go]: We should move this up per https://github.com/kubernetes/kubernetes/pull/39404#discussion_r110540666

**Comment by shiywang** [comment]
@adohe is there any progress on this ?

**Comment by shiywang** [comment]
@adohe is there any progress on this ?

File: pkg/kubectl/cmd/util/editor/editoptions.go (hunk 4 above)
Comment (adohe) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: comment addressed.

File: pkg/kubectl/cmd/util/editor/editoptions.go (hunk 6 above)
Comment (adohe) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: Done.

File: pkg/kubectl/cmd/util/editor/editoptions.go (hunk 7 above)
Comment (adohe) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: Done.

File: pkg/kubectl/cmd/util/editor/editoptions.go (hunk 6 above)
Comment (adohe) [inline pkg/kubectl/cmd/util/editor/editoptions.go]: Done.

**Comment by adohe** [comment]
rebased and comments addressed, ptal.

**Comment by mengqiy** [comment]
Review status: 0 of 9 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/kubectl/cmd/util/helpers.go, line 430 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39404#-Ki3LZ5eorKAMNC48iol:-Ki3LZ5fUDhukPQfnqTd:buiit5g) ([raw file](https://github.com/kubernetes/kubernetes/blob/5ed1ab66a293bebc03219d1338804a2131ba4271/pkg/kubectl/cmd/util/helpers.go#L430)):*
> ```Go
> 
> func AddApplyAnnotationVarFlags(cmd *cobra.Command, applyAnnotation *bool) {
> 	cmd.Flags().BoolVar(applyAnnotation, ApplyAnnotationsFlag, false, "If true, the configuration of current object will be saved in its annotation. This is useful when you want to perform kubectl apply on this object in the future.")
> ```

Flag help text inconsistent.

---

*[pkg/kubectl/cmd/util/editor/editoptions.go, line 349 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39404#-KhKpZHGmYVmvaznXohd:-Ki3NqX2_7_viCJDvSzw:byl7x06) ([raw file](https://github.com/kubernetes/kubernetes/blob/7fc7847b0623093e6eef73e818f7adf5dc2575c0/pkg/kubectl/cmd/util/editor/editoptions.go#L349)):*
<details><summary><i>Previously, shiywang (Shiyang Wang) wrote…</i></summary><blockquote>

I'm not sure if this is right, previously if  specify a invalid ouput, it will prompt  "The flag 'output' must be one of yaml|json" should we keep error output compatible ?
</blockquote></details>

This is not identical to before.

Before is
switch format {
case "json":
    // json printer
case "yaml", "": // specify yaml or unspecified
    // yaml printer
default:
    // error
}

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39404)*

**Comment by mengqiy** [comment]
I'm OK to merge this as is and let @shiywang to address the comments left in a followup PR.
This PR is blocking @shiywang's PR https://github.com/kubernetes/kubernetes/pull/42256.
It seems @adohe don't have bandwidth on this.

**Comment by kargakis** [comment]
sgtm

**Comment by mengqiy** [comment]
/lgtm
And @shiywang will address the rest.

**Comment by mengqiy** [comment]
Created https://github.com/kubernetes/kubernetes/issues/44905 to make sure the comments left will be addressed.
